File: TeamCode/src/main/java/brickbot/old/quickstart/commandbase/TimedCommand.java
package brickbot.old.quickstart.commandbase;


import androidx.annotation.NonNull;

import com.qualcomm.robotcore.util.ElapsedTime;

/**
 * Usage:
 * Allows any function and will return true if the timeout is reached
 * or the function returns true(unless it is not a boolean lambda)
 * new TimedCommand(() -> class.method(), timeoutSeconds);
 */
public class TimedCommand extends Command {
    private ElapsedTime timer;
    private final double timeoutSeconds;
    private final LambdaFunction<?> function;

    /**
     * Allows any function and will return true if the timeout is reached
     * @param function The function to run
     * @param timeoutSeconds The amount of time to wait
     */
    public TimedCommand(LambdaFunction<?> function, double timeoutSeconds) {
        this.function = function;
        this.timeoutSeconds = timeoutSeconds;
    }

    /**
     * Allows any function and will return true if the timeout is reached
     * @param commandName The name of the command
     * @param function The function to run
     * @param timeoutSeconds The amount of time to wait
     */
    public TimedCommand(String commandName, LambdaFunction<?> function, double timeoutSeconds) {
        super(commandName);
        this.function = function;
        this.timeoutSeconds = timeoutSeconds;
    }

    /**
     * Allows any function and will return true if the timeout is reached
     * @param function The function to run
     * @param timeoutSeconds The amount of time to wait
     */
    public TimedCommand(Runnable function, double timeoutSeconds) {
        this.function = () -> {
            function.run();
            return null;
        };
        this.timeoutSeconds = timeoutSeconds;
    }

    /**
     * Allows any function and will return true if the timeout is reached
     * @param commandName The name of the command
     * @param function The function to run
     * @param timeoutSeconds The amount of time to wait
     */
    public TimedCommand(String commandName, Runnable function, double timeoutSeconds) {
        super(commandName);
        this.function = () -> {
            function.run();
            return null;
        };
        this.timeoutSeconds = timeoutSeconds;
    }

    @Override
    public boolean run() {
        if(timer == null)
            timer = new ElapsedTime();

        boolean value = function.run() != null && (boolean) function.run();

        //TODO: test if this works
        return timer.seconds() >= timeoutSeconds || value;
    }

    @NonNull
    @Override
    public Command clone() {
        return new TimedCommand(getCommandName(), function, timeoutSeconds);
    }
}

File: TeamCode/src/main/kotlin/brickbot/quickstart/commandbase/TimedCommand.kt
package brickbot.quickstart.commandbase

import com.qualcomm.robotcore.util.ElapsedTime
import java.util.function.BooleanSupplier

class TimedCommand @JvmOverloads constructor(
    commandName: String = "",
    private val function: BooleanSupplier,
    private val timeoutMs: Long
): Command(commandName) {
    private lateinit var timer: ElapsedTime

    @JvmOverloads
    constructor(commandName: String = "", function: Runnable, timeoutMs: Long): this(
        commandName,
        BooleanSupplier {
            function.run()
            false
        },
        timeoutMs
    )

    override fun run(): Boolean {
        if (!::timer.isInitialized) {
            timer = ElapsedTime()
            timer.reset()
        }

        return function.asBoolean || (timer.milliseconds() > timeoutMs)
    }

    override fun clone(): Command {
        return TimedCommand(commandName, function, timeoutMs)
    }
}